pkg/runcfanotify/runcfanotify.go

```diff
@@ -240,14 +243,41 @@ func (n *RuncNotifier) AddWatchContainerTermination(containerID string, containe
 
 	n.containers[containerID].pidfd = int(pidfd)
 
+	n.awakeWatchContainersTermination()
+
 	return nil
 }
 
+// awakeWatchContainersTermination writes to the pipe so
+// watchContainersTermination() unblocks waiting on poll() and can
+// update the list of the watched containers or return if the
+// runcnotifier instance was closed
+func (n *RuncNotifier) awakeWatchContainersTermination() {
+	if n.pipeFdW != -1 {
+		f := []byte{0}
+		unix.Write(n.pipeFdW, f)
+	}
+}
+
 // watchContainerTermination waits until the container terminates using
 // pidfd_open (Linux >= 5.3), then sends a notification.
 func (n *RuncNotifier) watchContainersTermination() {
 	defer n.wg.Done()
 
+	// create a pipe to use the self-pipe trick to "unblock" poll
+	pipeFds := make([]int, 2)
+
+	if err := unix.Pipe2(pipeFds, unix.O_NONBLOCK); err != nil {
```

Comment:
It bothers me a bit that the pipe is created in the go routine because awakeWatchContainersTermination() might be called before the pipe is created. I know that you test `if n.pipeFdW != -1` but `pipeFdW` is not protected by a mutex so I am not sure if the compiler would retain the instruction ordering in this function.

What about creating the pipe in the caller NewRuncNotifier() before creating the go routine:

```
	n := &RuncNotifier{
		callback:   callback,
		containers: make(map[string]*runcContainer),
                pipeFds: []int{-1, -1}
...
	if pidfdOpenAvailable {
		// XXX: call Pipe2() here
		go n.watchContainersTermination()
```

and closing the `pipeFds` in RuncNotifier.Close() (after wg.Wait) instead of using a defer in the go routine.

Reply:
I think that would also fix the unit test `TestDNS` because at the time we call `initialFdList := currentFdList`, the pipes would have been created for sure. With the current code, the pipes might or might not be created, depending on how fast the go routine starts.

Reply:
Yes, that's much better. It also helps to handle the return error from Pipe2(). Updated the PR.